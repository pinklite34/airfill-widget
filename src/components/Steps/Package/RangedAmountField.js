import React from 'react';
import styled from 'styled-components';
import Field from '../../UI/Field';

const Input = styled.input`
  min-width: 8em;
  margin-right: 8px;
  padding-right: 3em;
  padding: 8px;
  appearance: textfield;

  &::-webkit-outer-spin-button,
  &::-webkit-inner-spin-button {
    -webkit-appearance: none;
  }
`
const Label = styled.label`
  width: 3em;
  position: relative;
  left: -3.25em;
  margin-right: -2.25em;
  background-color: #fff;
  color: #777;
`
const Price = styled.strong`
  color: #5ab76b;
`

const RangedAmountField = ({
  amount, range, currency, billingCurrency, onChange, onBlur
}) => {
  let error;

  amount = Number(amount);
  let currentPrice = amount * range.userPriceRate;

  // Round decimal ranges since we do not accept them atm
  const min = Math.ceil(range.min);
  const max = Math.floor(range.max);

  // BTC
  if (billingCurrency === 'XBT') {
    currentPrice = Math.ceil(currentPrice / 100) / 1000000;
  } else { // USD or EUR
    currentPrice = currentPrice.toFixed(2);
  }

  const displayedCurrency = billingCurrency === 'XBT' ? 'BTC' : billingCurrency;
  const hint = (
    <span className="">
      You pay <Price className="amount-ranged-btc-price">{currentPrice} {displayedCurrency}</Price>.
      Min: <strong>{min} {currency}</strong>, max: <strong>{max} {currency}</strong>.
    </span>
  );

  // Value not an integer
  if ((amount % 1) !== 0) {
    error = (
      <span>
        <strong>{amount}</strong> is not an accepted amount. Please
        choose <strong>{Math.floor(amount)} {currency}</strong> or
        <strong> {Math.ceil(amount)} {currency}</strong>.
      </span>
    );
  }
  // Value out of range
  else if (amount < min || amount > max) {
    error = (
      <span>
        Please choose an amount between <strong>{min} {currency}
        </strong> and <strong>{max} {currency}</strong>.
      </span>
    );
  }
  // Value not a correct step
  else if (range.step && (amount % range.step) !== 0) {
    const lowerStep = amount - (amount % range.step);
    error = (
      <span>
        <strong>{amount}</strong> is not an accepted amount. Please
        choose <strong>{lowerStep} {currency}</strong> or
        <strong>{(lowerStep + range.step)} {currency}</strong>.
      </span>
    );
  }

  return (
    <Field
      label="...or enter a custom amount"
      htmlFor="custom_amount"
      hint={hint}
      error={error}
    >
      <div>
        <Input
          type="number"
          id="custom_amount"
          name="custom_amount"
          className="amount-ranged-input"
          min={min}
          max={max}
          step={range.step}
          value={String(amount)}
          placeholder="e.g. 200"
          onChange={(e) => onChange(parseInt(e.target.value, 10))}
          onBlur={(e) => onBlur(parseInt(e.target.value, 10))}
        />
        <Label htmlFor="custom_amount">{currency}</Label>
      </div>
    </Field>
  );
};

export default RangedAmountField;
