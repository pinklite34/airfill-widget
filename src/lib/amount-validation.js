import { getDisplayName } from './currency-helpers';

// Handle XBT
export const getCostForAmount = (conversionRate, amount) =>
  Number(amount * conversionRate).toFixed(2);

const pickMiddlePackage = ({ packages }) => {
  const middle = Math.round((packages.length - 1) * 0.6);
  const pkg = packages[middle];
  return pkg ? String(pkg.value) : null;
};

const pickAffordablePackage = ({ packages, amount, currency, maxCost }) => {
  const costKey = getDisplayName(currency).toLowerCase() + 'Price';

  const isAffordable = pkg => pkg[costKey] <= maxCost;

  // Make sure user can afford the selected package
  const selectedPackage = packages.find(pkg => String(pkg.value) === amount);
  if (selectedPackage && isAffordable(selectedPackage)) {
    return amount;
  }

  // Otherwise try to pick the highest value package the user can afford
  const highestAffordablePackage = packages.filter(isAffordable).pop();

  if (highestAffordablePackage) {
    return Number(highestAffordablePackage.value);
  }

  // If there are no packages the user can afford, return the amount as is
  return amount;
};

export const selectValidPackage = ({ packages, maxCost, currency, amount }) => {
  // If no amount is selected, pick a package in the middle
  amount = amount ? String(amount) : pickMiddlePackage({ packages });

  // Make sure user can afford the package
  return pickAffordablePackage({ packages, amount, maxCost, currency });
};

const rangedCostForAmount = (conversionRate, currency, amount) => {
  let amountCost = amount * conversionRate;
  // BTC
  if (currency === 'XBT') {
    return Math.ceil(amountCost / 100) / 1000000;
  } else {
    // USD or EUR
    return Number(amountCost.toFixed(2));
  }
};

const selectValidRangedAmount = ({
  amount,
  maxCost,
  costConversionRate,
  currency
}) => {
  const selectedAmountCost = rangedCostForAmount(
    costConversionRate,
    currency,
    amount
  );

  if (selectedAmountCost <= maxCost) {
    return String(amount); // Return amount as is for ranged operators
  } else {
    if (currency === 'XBT') {
      const amountForMaxCost = Math.floor(
        maxCost * 100000000 / costConversionRate
      );
      return String(amountForMaxCost); // Return the maximum amount allowed
    } else {
      return String(maxCost / costConversionRate);
    }
  }
};

export const selectValidAmount = args => {
  const ranged = args.ranged;
  const validPackageAmount = selectValidPackage(args);

  if (ranged) {
    return selectValidRangedAmount({
      ...args,
      amount: args.amount || validPackageAmount
    });
  } else {
    return validPackageAmount;
  }
};

