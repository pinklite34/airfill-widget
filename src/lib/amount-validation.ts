import { Amount, BillingCurrency, Package, RangeProp } from '../types';

import { getDisplayName } from './currency-helpers';

// if package is affordable with current user balance
// is affordable if no balance is set
export function isAffordable(
  pkg: Package,
  billingCurrency: BillingCurrency,
  balance: Amount
) {
  const costKey = getDisplayName(billingCurrency).toLowerCase() + 'Price';

  return !balance || pkg[costKey] <= balance;
}

export function selectValidAmount(
  packages: Package[],
  billingCurrency: BillingCurrency,
  balance?: Amount,
  range?: RangeProp
) {
  if (range) {
    const correctMax =
      billingCurrency === 'XBT'
        ? ((balance as number) * 100000000) / range.userPriceRate
        : (balance as number) / range.userPriceRate;

    if (correctMax < range.min) {
      return 0;
    } else if (correctMax < range.max) {
      return correctMax.toFixed();
    } else {
      return range.max;
    }
  } else {
    const costKey = getDisplayName(billingCurrency).toLowerCase() + 'Price';

    const f = packages
      .sort((a, b) => (a[costKey] < b[costKey] ? 1 : -1))
      .filter(pkg => isAffordable(pkg, billingCurrency, balance));

    // pick highest affordable or most expensive, if no affordable
    return f.length > 0 ? f[0].value : packages[0].value;
  }
}
