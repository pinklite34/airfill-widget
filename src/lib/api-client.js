import ponyFetch from 'fetch-ponyfill';
const { fetch: _fetch } = ponyFetch();

export function encodeQueryString(obj) {
  const queryString =
    typeof obj === 'object' &&
    !!obj &&
    Object.keys(obj)
      .filter(k => obj[k])
      .map(k => encodeURIComponent(k) + '=' + encodeURIComponent(obj[k]))
      .join('&');

  return queryString ? '?' + queryString : '';
}

export const createClient = (conf = {}) => {
  let baseUrl = conf.baseUrl || '/api';
  let { username, password, defaultHeaders, token } = conf;

  const doFetch = (uri, options) => {
    let url = baseUrl + uri;

    // Set default headers
    options.headers = {
      Accept: 'application/json',
      ...defaultHeaders,
      ...options.headers,
    };

    // Build query string if needed
    if (options.query) {
      url = url + encodeQueryString(options.query);
    }

    // Add body data
    if (options.body) {
      options.headers = {
        'Content-Type': 'application/json',
        ...options.headers,
      };

      if (typeof options.body !== 'string') {
        options.body = JSON.stringify(options.body);
      }
    }

    options.method = options.method || (options.body ? 'post' : 'get');

    let user = options.username || username;
    let pass = options.password || password;

    // Add auth if provided
    if (user && pass) {
      options.headers = {
        Authorization: 'Basic ' + btoa(user + ':' + pass),
        ...options.headers,
      };
    } else if (token) {
      options.headers = {
        Authorization: 'Token ' + token,
        ...options.headers,
      };
    }

    return _fetch(url, options)
      .then(response => {
        if (response.status >= 200 && response.status < 300) {
          return response.json();
        }
        throw response;
      })
      .catch(error => {
        if (typeof error && error.json === 'function') {
          return error
            .json()
            .then(json => {
              error.json = json;

              if (json.error) {
                return json.error.message;
              }
              if (json.message) {
                return json.message;
              }

              throw Error('No error information in JSON response');
            })
            .catch(() => {
              if (error.statusText) {
                return error.statusText;
              } else {
                return 'Server returned status code ' + error.status;
              }
            })
            .then(message => {
              const err = new Error(message);
              err.response = error;
              throw err;
            });
        } else {
          console.error(error.message || error);
        }
      });
  };

  const configure = conf => {
    baseUrl = 'baseUrl' in conf ? conf.baseUrl : baseUrl;
    username = 'username' in conf ? conf.username : username;
    password = 'password' in conf ? conf.password : password;
    token = 'token' in conf ? conf.token : token;
    defaultHeaders =
      'defaultHeaders' in conf ? conf.defaultHeaders : defaultHeaders;
  };

  const createHref = (uri, queryParams) => {
    return baseUrl + (uri || '') + encodeQueryString(queryParams);
  };

  return { fetch: doFetch, configure, createHref };
};

export const client = createClient();

export function fetch(uri, options) {
  return client.fetch(uri, options);
}
