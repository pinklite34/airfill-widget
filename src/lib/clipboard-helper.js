export default text => {
  const element = document.createElement('textarea');
  element.textContent = text;
  document.body.appendChild(element);
  element.select();

  try {
    document.execCommand('copy');
  } catch (e) {}

  document.body.removeChild(element);
  window.alert('Copied to clipboard!');
};
