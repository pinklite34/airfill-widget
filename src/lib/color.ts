import * as Color from 'color';

export function colorToString(color) {
  return Color(color)
    .hsl()
    .string();
}

export function colorDarken(color, amount) {
  return colorToString(Color(color).darken(amount));
}

export function colorLighten(color, amount) {
  return colorToString(Color(color).lighten(amount));
}
