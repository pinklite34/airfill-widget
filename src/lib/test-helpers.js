import renderer from 'react-test-renderer';

export const matchRenderWithSnapshot = jsx => {
  const tree = renderer.create(jsx).toJSON();
  expect(tree).toMatchSnapshot();
}
