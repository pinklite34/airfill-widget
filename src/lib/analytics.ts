import { Order, Product, RootState } from '../types';

import { getPlatform, getSource, isMobileApp } from '../lib/globals';

export function productPropertiesForOrder(order?: Order) {
  return order
    ? {
        product_id: order.operator,
        sku: order.operator,
        name: order.itemDesc,
        price: order.eurPrice,
        quantity: 1,
        category: order.operatorType,
        currency: 'EUR',
      }
    : {};
}

export function eventPropertiesForOrder(order?: Order) {
  return order
    ? {
        ...order,
        order_id: order.id,
        affiliation: 'bitrefill',
        value: order.eurPrice,
        revenue: order.eurPrice,
        label: order.paymentMethod,
        currency: 'EUR',
        products: [productPropertiesForOrder(order)],
      }
    : null;
}

export function eventPropertiesForProduct(product?: Product) {
  return product
    ? {
        ...product,
        product_id: product.slug,
        sku: product.slug,
        category: product.type,
        name,
        image_url: product.logoImage,
      }
    : null;
}

export function createEvent(
  state: RootState,
  eventType: string,
  payload
): object {
  payload = payload || { properties: {} };

  const source = getSource();
  const platform = getPlatform();

  const additionalKey = payload.traits ? 'traits' : 'properties';
  const additionalProps = {
    category: isMobileApp() ? platform : 'website',
    user_source: source,
    user_source_platform: platform,
  };

  return {
    eventType,
    eventPayload: {
      ...payload,
      [additionalKey]: {
        ...additionalProps,
        ...payload[additionalKey],
      },
    },
  };
}
