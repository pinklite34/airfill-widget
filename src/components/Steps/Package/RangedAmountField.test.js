import React from 'react';
import { matchRenderWithSnapshot } from '../../../lib/test-helpers';

import RangedAmountField from './RangedAmountField.js';

describe('RangedAmountField', () => {
  it('renders with the min/max rounded up/down', () => {
    matchRenderWithSnapshot(<RangedAmountField
      amount={18}
      range={{min: 4.28, max: 50.28, userPriceRate: 1}}
      currency={'EUR'}
      billingCurrency={'XBT'}
    />);
  });
});
