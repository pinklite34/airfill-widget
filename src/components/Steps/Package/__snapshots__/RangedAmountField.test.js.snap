// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangedAmountField renders with the min/max rounded up/down 1`] = `
<div
  className="hulnrW"
>
  <label
    className="refill-field-label"
    htmlFor="custom_amount"
  >
    ...or enter a custom amount
  </label>
  <div
    className="refill-field"
  >
    <div>
      <input
        className="amount-ranged-input jsqEwQ"
        id="custom_amount"
        max={50}
        min={5}
        name="custom_amount"
        onChange={[Function]}
        placeholder="e.g. 200"
        step={undefined}
        type="number"
        value={18}
      />
      <label
        className="fEuSWJ"
        htmlFor="custom_amount"
      >
        EUR
      </label>
    </div>
  </div>
  <p
    className="refill-field-hint"
  >
    <span
      className=""
    >
      You pay 
      <strong
        className="amount-ranged-btc-price fcgrxX"
      >
        0.000001
         
        BTC
      </strong>
      . Min: 
      <strong>
        5
         
        EUR
      </strong>
      , max: 
      <strong>
        50
         
        EUR
      </strong>
      .
    </span>
  </p>
</div>
`;
